[v1][sampler] Inplace logprobs comparison to get the token rank

houseroad opened the pull request:
Summary: Original implementation is inefficient, since we will create a copy of the original tensor. So create a triton kernel to get this done in place.

Test Plan

Launch the server with warm up requests

Before:
OOM

After:
No more OOM

Differential Revision: D78606604

Jialin commented on vllm/v1/sample/ops/logprobs.py:
```diff
@@ -0,0 +1,107 @@
+
+    # 6. Perform the comparison and sum the result within this block.
+    # `x_block > value` creates a boolean vector, and tl.sum treats True as 1.
+    local_count = tl.sum(x_block > value)
```
IIUC, this PR is for HBM usage reduction?
- Originally, x >= values would create a new bool tensor before count True against it, so the HBM overhead would be x.nbytes.
- In the new implementation, x_block > value would still create boolean vector before accumulatiing with tl.sum. However, blocks are not executed at the exact same time, so the HBM usage would become max_concurrent_block * x_block.nbytes

Is my understand accurate?

I kinda feel we're having quite a lot usage similar to (x >= y).sum(), should we introduce a new pytorch API to optimize this, in general.

Jialin commented:
Maybe add unittest to confirm the correctness?

houseroad committed:
Inplace logprobs comparison to get the token rank (#21283)

Summary:
Pull Request resolved: https://github.com/vllm-project/vllm/pull/21283

Original implementation is inefficient, since we will create a copy of the original tensor. So create a triton kernel to get this done in place.

Test Plan:
```
VLLM_GPU_MEMORY_UTILIZATION=0.8 \
VLLM_USE_FLASHINFER_SAMPLER=0 \
CUDA_VISIBLE_DEVICES=0 \
VLLM_USE_V1=1 \
SAFETENSORS_FAST_GPU=1 \
buck2 run mode/opt mode/inplace \
  -c fbcode.platform010_cuda_version=12.8 \
  -c fbcode.enable_vllm=true \
  -c fbcode.enable_gpu_sections=true \
  -c fbcode.nvcc_arch=h100a //smart/inference_platform_sp/llm_predictor_gpu:service -- \
  --max_seq_len=8192 \
  --max_batch_size=64 \
  --model_mf_bucket=fair_llms_prod \
  --model_mf_path=tree/894459467/0 \
  --allow_custom_stop_tokens \
  --vllm_engine --local_cache_dir=/tmp/llama4-planner \
  --try_local_cache \
  --thrift_server_port 12345 \
  --model_parallel_size=1 \
  --thrift_queue_timeout_ms=500 \
  --guided_decode \
  --token_logprobs
```

No more OOM

Rollback Plan:

Reviewed By: JialinOuyang-Meta

Differential Revision: D78606604

Signed-off-by: Lu Fang <lufang@fb.com>

houseroad commented:
Actually, switched to torch.compile, no more OOM. Thanks to @yinghai for the suggestion.

houseroad commented:
The overhead should be fine, since it's only a few ops, usually this is what torch.compile is really great at.
